text/0000-sandbox-environment.md

```diff
@@ -0,0 +1,201 @@
+   deployment environment (such as when cross-compiling).
+
+This RFC proposes a way to precisely control the environment visible to the compile-time macros, while defaulting to the
+current behaviour.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Rust implements the `env!()` and `option_env!()` macros to access the process environment variables at compilation time.
+`rustc` supports a number of command-line options to control the environment visible to the compiling code.
+
+By default all environment variables are available with their value taken from the environment. There are several
+additional controls to control the logical environment accessed by `env!()`/`option_env!()`:
+- only allow access to a specific whitelist of variables
+- override specific variables to other values
+- add new variables without them being present in the environment
+
+These options are:
+- `--env-whitelist REGEX` - match the REGEX against all existing process environment
+  variables and allow them to be seen. This overrides `--env-remove-all`. The regex is matched against the entire variable name
+  (that is, it is anchored).
+- `--env-blacklist REGEX` - match the REGEX against the environment and remove those variables from the logical environment; this
+  is equivalent to unsetting them from the Rust code's perspective.
+- `--env-set VAR=VALUE` - set the logical value of an environment variable. This will override the value if it already exists
+  in the process environment, or create a new logical environment variable.
+
+These options are processed in order. For example:
+```
+rustc --env-blacklist '.*' --env-whitelist 'CARGO_.*' --env-set HOME=/home/system [...]
+```
+will clean all environment variables from the logical environment. It then allows access to all Cargo-set variables, and overrides
+the value of `$HOME`.
+
+Note that these options act on the logical environment, so:
+```
+rustc --env-set FOO=BAR --env-blacklist FOO
+```
+will leave `FOO` unset.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The implementation of this RFC introduces the notion of a logical environment which is accessed by the `env!`/`option_env!` macros,
+distinct from the actual process environment. By default they are the same, but the additions of the `--env-whitelist`,
+`--env-blacklist` and `--env-set` options allow the logical environment to be tailored as desired.
+
+## Processing of the options
+
+The `--env-` options are processed in the order they appear on the command-line, left to right. The logical environment is
+initialized from the process environment. Then each each `--env` option is processed in turn, as it appears, to update the logical
+environment. Specifically:
+
+- `--env-whitelist REGEX` - Any name which doesn't match the REGEX is removed from the logical environment,
```

Comment:
Is the regular expression needed? Explicitly listing all individual variables seems not unreasonable.
Another question: does `TEST` match `NOT_TEST`?

Reply:
I chose regex for things like allowing all CARGO_ cars without having to enumerate all the features or whatever else cargo adds.

I mentioned that the regex is anchored so it implicitly has ^$ around it - so no TEST wouldn't match NOT_TEST.